### DISCUSSION TITLE
Feature request: Make dev server handle HEAD and byte range requests

### DISCUSSION THREAD

**Discussion by jordemort** [withastro/roadmap#309] [Proposal]
I'm developing a client-side search feature for my [blog](https://jordemort.dev/) using https://github.com/phiresky/sql.js-httpvfs

Right now I am building my search index in a custom Astro integration, that runs after `astro:build:done`. This works well, but it means my search index is not available during development.

I would like to switch this to using a [file route](https://docs.astro.build/en/core-concepts/astro-pages/#file-routes) but I ran into two problems:

- sql.js-httpvfs does a `HEAD` request on the database file before trying to use it; the dev server was returning a 404 error instead of the header of the response. I was able to get around this by exporting a `head` function from my file route, but...
- Astro's dev server does not implement [byte-range requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests), which are required for `sql.js-httpvfs` to work. I was not able to work around this; while my `get` and `head` functions do receive a `Request` object as their first parameter, this `Request` object does not contain any of the headers from the client, so I am unable to emulate support for byte range requests.

I realize that this is probably a very niche thing, but it has made developing my search feature very painful. I can't use the dev server; I have to build my site, serve it up with a local NGINX, and debug it from there. The compile/run/debug cycle I've gotten myself into with this reminds me of working on a C++ program.

Here are the things that I wish Astro would do for me:

- If a file route does not implement `head`, provide a default implementation that just calls `get` and then chops off the content.
- Support byte range requests:
  - A `HEAD` should:
    1. Generate the entire response content
    2. Calculate the length of the body in bytes and add a `Content-Length` header
    3. Discard the body
    4. Add an `Accept-Ranges: bytes` header
    5. Serve the headers
  - A `GET` should behave exactly as before if there is not a `Range: bytes` header.
  - A `GET` with a `Range: bytes` header should:
    1. Generate the entire response content
    2. Discard any of body outside of the range specified by the header
    3. Return response code 206 ("partial content"), with a `Content-Type` of `multipart/byteranges` and the response with the modified body nested inside of it

**Comment by vinnichase** [comment]
Hey I'm +1 on the defaults for HEAD and GET requests.

For your custom integration though... Have you tried these hooks to keep your index up-to-date during development:
```TS
hooks: {
    'astro:build:done': ({ logger }) => {
        doStuff(logger);
    },
    'astro:server:start': ({ logger }) => {
        doStuff(logger);
    },
    'astro:server:setup': ({ logger, server }) => {
        server.watcher.on('change', file => {
            doStuff(logger);
        });
    },
},
```
